### PR TITLE
Srtf Scheduling

### LINKED ISSUE #1790: Shortest remaining time first cpu scheduler
Being a participant of gssoc 2020  I would like to work on this issue.I would like to implement srtf(preemptive) in java.

-Vishakha Narang

### PR DESCRIPTION
**Pull request by vishakha2108**
Fixes #1790 

#### Checklist:
- [* ] 4 space indentation.
- [ *] Coding conventions are followed.
- [* ] Input is taken dynamically.
- [* ] Sample Input / Output is added at the end of file.
- [ *] Logic Documentation (Comments).
- [* ] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- added code for SRTF scheduling
- 

#### Languages Used:
- java
- 

#### Files Added:
- 
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by asha15** [comment]
could you provide a link to your issue ?

**Comment by vishakha2108** [comment]
https://github.com/jainaman224/Algo_Ds_Notes/pull/2250#

**Comment by asha15** [comment]
No you are giving the earliest PR number. I'm asking your issue number where the mentor assigned it to you. can i know it?

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
+
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:1]: remove this line

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+{
+    static Scanner in = new Scanner(System.in);//for input
+    float At = 0;//parameters of process
+    float Bt = 0;
+    float Ct = 0;
+    float Tt = 0;
+    float Wt = 0;
+    float Timeleft ;//only used for srtf
+
+    Process(float a , float b)// constructor
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:15]: remove the space before the comma

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    {
+        At = a;
+	Bt = b;
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:18]: Correct the indentation

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    public static float calTt(Process p[], int n)//tt = ct-at and avg tt
+    {
+        float avg = 0;//avg
+	for(int i = 0;i<n;i++)
+	{
+            p[i].Tt = p[i].Ct-p[i].At;
+	    avg = avg+p[i].Tt;
+	}
+	avg = avg/n;
+	return avg;
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:30]: correct the indentation

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    public static float calWt(Process p[], int n)//wt = tt-bt and avg wt
+    {
+        float avg = 0;//avg
+	for(int i = 0;i<n;i++)
+	{
+            p[i].Wt = p[i].Tt-p[i].Bt;
+	    avg = avg+p[i].Wt;
+	}
+	avg = avg/n;//avg wt
+	return avg;
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:42]: correct the indentation

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+class SRTF
+{	
+    float Awt;
+    float Att;
+    float sumBT;
+    public SRTF()
+    {
+        sumBT = 0;
+    }
+    static Scanner in = new Scanner(System.in);
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:55]: keep a line space after this

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    public void Accept(Process p[],int n)//accept func
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:56]: keep a space after comma

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    {
+        for(int i = 0;i<n;i++)
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:58]: keep a space after operator

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	{
+            System.out.println("Enter arrival time for process P"+i);//at
+            float a = in.nextFloat();
+	    System.out.println("Enter burst time for process P"+i);//bt
+	    float b = in.nextFloat();
+	    p[i] = new Process(a,b);
+	    p[i].Timeleft = p[i].Bt;//used to find shortest process remaining
+	    sumBT= sumBT+p[i].Bt;//used for cal ct in calcct function
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:66]: correct the indentation

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+	 }
+    }
+   
+    public void CalcCT(Process p[],int n)
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:70]: keep a space after comma

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+	    {
+                p[min].Ct = i+1;// +1 because it will complete in next cycle
+	        p[min].Timeleft = 32767;
+	     }
+	     System.out.print("P"+min+"| ");//gantt chart
+	}    
+	Att = Process.calTt(p, n);//att
+	Awt = Process.calWt(p, n);//awt
+    }
+    public void Display(Process p[],int n)
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:95]: correct the indentation

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    { 
+        System.out.println();
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:97]: indentation

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    {
+        Scanner in = new Scanner(System.in);
+        System.out.println("SRTF algorithm");
+        int n;//no of processes
+        System.out.println("Enter no of process");//ip of process
+        n = in.nextInt();
+        Process P[] = new Process[n];
+	SRTF S  = new SRTF();//object of srtf
+	S.Accept(P,n);
+        S.CalcCT(P,n);//all calculations and gantt chart
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:123]: keep a space after comma

File: SRTF_Scheduling/SRTF_Scheduling.java (hunk 7 above)
Comment (asha15) [reply SRTF_Scheduling/SRTF_Scheduling.java:56]: please resolve this

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    float sumBT;
+    public SRTF()
+    {
+        sumBT = 0;
+    }
+    static Scanner in = new Scanner(System.in);
+    
+    public void Accept(Process p[],int n)//accept func
+    {
+        for(int i = 0;i< n;i++)
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:58]: keep a space after ;

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        {
+            System.out.println("Enter arrival time for process P"+i);//at
+            float a = in.nextFloat();
+            System.out.println("Enter burst time for process P"+i);//bt
+            float b = in.nextFloat();
+            p[i] = new Process(a,b);
+            p[i].Timeleft = p[i].Bt;//used to find shortest process remaining
+            sumBT= sumBT+p[i].Bt;//used for cal ct in calcct function
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:66]: keep a space before and after operator

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        }
+    }
+   
+    public void CalcCT(Process p[], int n)
+    {
+        System.out.println();
+        System.out.println();
+        int min = 0;
+        for(int i = 0;i< sumBT;i++)
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:75]: keep space after operator

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+                    min = j;//  process index with min bt
+                }
+            }
+            p[min].Timeleft--;
+            if(p[min].Timeleft==0)//setting ct when execution is done
+            {
+                p[min].Ct = i+1;// +1 because it will complete in next cycle
+                p[min].Timeleft = 32767;
+            }
+            System.out.print("P"+min+"| ");//gantt chart
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:90]: please keep space around operators

File: SRTF_Scheduling/SRTF_Scheduling.java [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+    { 
+        System.out.println();
+        System.out.println("process  Arrival Time  Burst Time  Completion Time  Turnover Time  Waiting Time");//table
+        for(int i = 0;i< n;i++)
+        {
+            System.out.println("P"+i+"\t "+p[i].At+"\t\t"+p[i].Bt+"\t\t"+p[i].Ct+"\t\t"+p[i].Tt+"\t\t"+p[i].Wt);   
+        }
+        System.out.println();
+        System.out.println("Average waiting time  is "+Awt);//other data to be displayed
+        System.out.println("Average turnover time  is "+Att);
```
Comment (asha15) [inline SRTF_Scheduling/SRTF_Scheduling.java:106]: keep space around operator

**Pull request merged**